config/bash/lxc.in

```diff
@@ -1,111 +1,1031 @@
+            ;;
+    esac
+
+    $split && return
+
+    if [[ $cur == -* ]]; then
+        COMPREPLY=( $( compgen -W '$( _parse_help "$1" )' -- "$cur" ) )
+        [[ ${COMPREPLY-} == *= ]] && compopt -o nospace
+        return
+    fi
+    _lxc_append_name
+}
+complete -F _lxc_create lxc-create
+
+_lxc_destroy() {
+    local cur prev words cword split
+    COMPREPLY=()
+    _init_completion -s -n : || return
 
-    complete -o default -F _lxc_generic_n lxc-attach
-    complete -o default -F _lxc_generic_n lxc-cgroup
-    complete -o default -F _lxc_generic_n lxc-console
-    complete -o default -F _lxc_generic_n lxc-destroy
-    complete -o default -F _lxc_generic_n lxc-device
-    complete -o default -F _lxc_generic_n lxc-execute
-    complete -o default -F _lxc_generic_n lxc-freeze
-    complete -o default -F _lxc_generic_n lxc-info
-    complete -o default -F _lxc_generic_n lxc-monitor
-    complete -o default -F _lxc_generic_n lxc-snapshot
-    complete -o default -F _lxc_generic_n lxc-start
-    complete -o default -F _lxc_generic_n lxc-stop
-    complete -o default -F _lxc_generic_n lxc-unfreeze
+    _lxc_common_opt || return
 
-    complete -o default -F _lxc_generic_ns lxc-wait
+    case $prev in
+        --name | -n )
+            _lxc_names
+            return
+            ;;
+        --rcfile )
+            _filedir
+            return
+            ;;
+        --force | -f | --snapshots | -s )
+            # Only flags.
+            ;;
+    esac
+
+    $split && return
+
+    if [[ $cur == -* ]]; then
+        COMPREPLY=( $( compgen -W '$( _parse_help "$1" )' -- "$cur" ) )
+        [[ ${COMPREPLY-} == *= ]] && compopt -o nospace
+        return
+    fi
+    _lxc_append_name
+}
+complete -F _lxc_destroy lxc-destroy
+
+_lxc_device() {
+    local cur prev words cword split
+    COMPREPLY=()
+    _init_completion -s -n : || return
+
+    case $prev in
+        -h )
+            return
+            ;;
+        --name | -n )
+            _lxc_names
+            return
+            ;;
+        add )
+            _available_interfaces
+            COMPREPLY+=( $(compgen -f -d -X "!*/?*" -- "${cur:-/dev/}") )
+            return
+            ;;
+    esac
+
+    $split && return
+
+    if [[ $cur == -* ]]; then
+        COMPREPLY=( $( compgen -W '$( _parse_help "$1" )' -- "$cur" ) )
+        [[ ${COMPREPLY-} == *= ]] && compopt -o nospace
+        return
+    fi
+    _lxc_append_name
+}
+complete -F _lxc_device lxc-device
+
+_lxc_execute() {
+    local cur prev words cword split
+    COMPREPLY=()
+    _init_completion -s -n : || return
+
+    _lxc_common_opt || return
+
+    case $prev in
+        --name | -n )
+            _lxc_names
+            return
+            ;;
+        --rcfile | -f )
+            _filedir
+            return
+            ;;
+        --define | -s )
+            # @TODO: list values from source code.
+            # tag=value
+            # https://github.com/lxc/lxc/blob/stable-4.0/src/lxc/confile.c#L178
```

Comment:
Ideally, this should be the output of a command from lxc. If it is not likely to change, though, we could just add the list manually like the templates and backing storages.

Reply:
That's a very long list and one that will pretty much always be partial as there are config keys that are dynamic (like the cgroup ones or environment keys). I personally tend to get annoyed when bash completion profiles try to return all the keys but then are missing a few.

My preference tends to be that either you can consistently list everything that's valid or you don't list anything.